File: Dart/1-50/15.dart
import 'dart:io';

const int mod = 1000000007;

int binPow(int a, int n) {
  if (n == 0) return 1;
  if (n % 2 == 0) {
    int x = binPow(a, n ~/ 2);
    return (x * x) % mod;
  }
  return (a * binPow(a, n - 1)) % mod;
}

void main() {
  List<String> input = stdin.readLineSync()!.split(' ');
  int n = int.parse(input[0]);
  int k = int.parse(input[1]);

  int a = binPow(k, n) - 1;
  int b = k - 1;
  int result = (a * binPow(b, mod - 2)) % mod;
  print(result);
}

File: Ruby/1-50/15.rb
MOD = 10**9 + 7

def binpow(a, n)
  return 1 if n == 0
  if n % 2 == 0
    x = binpow(a, n / 2)
    return (1 * x * x) % MOD
  end
  return (1 * a * binpow(a, n - 1)) % MOD
end

n, k = gets.chomp.split.map(&:to_i)
a = binpow(k, n) - 1
b = k - 1
result = (1 * a * binpow(b, MOD - 2)) % MOD
puts result
